### ISSUE TITLE
Project synthesis

### ISSUE THREAD

**Issue by amyjko** [wordplaydev/wordplay#387]
## What's the problem?

There are many examples we've built out that are compelling starting points. However, they are hard for students to imagine how to combine, and so the space of examples is sparse. 

## What's the design idea?

Generative LLM-based approaches, which would allow text descriptions of some behavior, generating a skeleton for students to start from, might be more compelling. There is much to do to help a student or teacher move from the skeleton, however, and most of these are open research questions:

* How can creators iterate on their program, either through their description or through code?
* How can creators use non-English languages to generate skeletons?
* How can generative AI techniques be used on a new programming language with very few examples? 

## Who benefits?

All creators starting a new project.

## Design specification

(This section should be included after a design proposal is ready and approved, and the `buildable` tag is added. This text can remain until then. Designers should add their proposal here, not in a comment).

**Comment by sdahl4** [comment]
I would like to be assigned to this issue to conduct a research literature review!

**Comment by amyjko** [comment]
Sure thing. Lots of literature in the past few years to study on this. Having a list of recent works with short summarize of what they teach us would be very valuable.

**Comment by sdahl4** [comment]
@amyjko Here is what I found after researching recent works about creating generative LLM tools from smaller datasets, as well as using those tools in non English languages: 

### Generative LLM-based approaches and Wordplay

Generative LLM-based programming tools become increasingly more used in the programming world, for example tools like github’s ai assistant 'co-pilot’ already having well over one million users. 
The questions posed with issue #387 is if could we utilize these generative LLM based tools in Wordplay. Could we have a feature where a user could generate code skeletons for projects based off their description?
Some questions that Dr Ko emphasizes when we think about how we could develop this type of tool for Wordplay is:
How would this tool interact with non-English based users?
Would a tool like a large language model be viable for a smaller data base like Wordplays?
This literature review looks at current works (mainly from 2024) and discusses how they relate to the posed questions above. I will also add my own personal view on this current research, and how we could utilize this research to decide whether to move forward in looking to create generative LLM based tools for Wordplay in the future, and if so, what tools would be useful for that process. 
I have categorized this review based off the two questions that arise from this issue, with an overview of  literature with my opinion after. 
### How can generative AI techniques be used on a new programming language with very few examples?

Ishfaq Hussain Rather, et al. “Breaking the Data Barrier: A Review of Deep Learning Techniques for Democratizing AI with Small Datasets.” Artificial Intelligence Review, vol. 57, no. 9, 2 Aug. 2024, https://doi.org/10.1007/s10462-024-10859-3.

This paper discusses the current way we look at language models they are require a lot data to work at their best, or at all.  This gives large companies with a plethora of resources a huge advantage over smaller organizations (like Wordplay)  that want to utilize learning machines. They suggest something called Deep Learning (DL) models, that require less data to create accurate learning models. 

The paper than goes on to list some techniques based in DL that can help solve this small dataset issue: 
First technique suggested is Data augmentation: where learning models with limited data face the challenge of ‘overfitting’ because of the limited exposure to different scenarios,  this paper suggest to solve this through augmentation. This is the generation of additional training samples by changing random parts of training data to create new samples from old ones. This could possibly be a tool we use to take an existing small dataset and augment it to expand the learning data. 
The paper goes on to mention some techniques for image based data learning which isn’t relevant to Wordplay, so I won’t go into describing those techniques here. 

I think data augmentation could potentially open some doors for Wordplay by creating a larger database of examples, but with the amount of data you need for LLMS even with these tools the Wordplay database of examples would be far smaller than the average database of examples used to create LLMS. 
I think it would be helpful to view data augmentation as something that could be a helpful tool in the process of creating an LLM, but not the solution to this data problem.

Mao, Yuren, et al. “A Survey on LoRA of Large Language Models.” ArXiv.org, 2024, arxiv.org/abs/2407.11046, https://doi.org/10.1007/s11704-024-40663-9 . Accessed 20 Nov. 2024.  ‌

This work looks at “Low-Rank Adaption” (LoRA) which is a tool to adapt pre trained language models to work on more specific tasks accurately. The main idea is with LoRA is that instead of having to train a new model which is expensive and time consuming, to update a pre existing model to learn how to do a new specific task, which in this case could be creating skeletons for WordPlay, and because is is just tweaking a pre existing LLM the dataset need would not nee to be as large. 
This works by adding “low-rank matrices’ to layers of the model to capture important knowledge specific to the task LoRA is teaching the model, think of it as adding a specific task ‘filter’ that the LLM has on top of it, instead of creating a new syststem, it adds on to the existing one. 

LoRA has potential for this data problem, but it might not be practical for generating actual code skeletons due to the need for high precision in programming. Even small errors can cause significant issues. However, it could be useful for generating code skeletons in the form of comments, guiding users on how to structure their program.

Yang, Y., Li, X., Zhou, Z., Song, S. L., Wu, J., Nie, L., & Ghanem, B. (2024). CorDA: Context-Oriented Decomposition Adaptation of Large Language Models for Task-Aware Parameter-Efficient Fine-tuning. Proceedings of the Conference on Neural Information Processing Systems (NeurIPS 2024).[ https://github.com/iboing/CorDA](https://github.com/iboing/CorDA)

I won’t go in to detail about this work, but I add it because this work, and several others like it, all work towards bridging the gaps left in performance by LLMs that use LoRA for specific tools, to show how their is a lot of resources and work going towards this idea of taking pre existing models and changing them to work for specific tasks, so perhaps this tool could become more of a viable option to use for Wordplay in the future. 

### How can creators use non-English languages to generate skeletons?
Wang, C., Li, Z., Gao, C., Wang, W., Peng, T., Huang, H., Deng, Y., Wang, S., & Lyu, M. R. (2024). Exploring multi-lingual bias of large code models in code generation (arXiv:2404.19368).[ https://arxiv.org/abs/2404.19368](https://arxiv.org/abs/2404.19368)
This paper looks at how other languages compared to English tend to have more discrepancies, since most training datasets are in English. 

The researcher investigated this issue by evaluating the code generation of nine different language models to analyze how  the  models respond in different languages (Chinese and English). They found that current models have a minimum decline of 13 percent in accuracy when they were requested to code something in Chinese vs in English. 
They also found that when comparing accuracy across languages in different coding languages, widely used languages like Python and Java had very few issues across languages but struggled with more intricate programming languages that are less used such as C++. 

Positively, the paper found that prompting models through translation helped mitigate the multi language bias. When the researchers used google translate to convert instructions from Chinese to English and than had the model run off of those translated instructions reduced errors in the code from 17.2 percent to 3.8. 
Maybe we could think about having some translation ‘assistance’ as a part of the process when using LLMs in languages that are different than the original training set data, which I believe is mostly English. 

Jiang, W., Gao, X., Zhai, J., Ma, S., Zhang, X., & Shen, C. (2024). From effectiveness to efficiency: Comparative evaluation of code generated by LCGMs for bilingual programming questions. arXiv.[ https://arxiv.org/abs/2406.00602](https://arxiv.org/abs/2406.00602)

This paper also looks in generation code in non english languages and comes to very similar conclusions as the one above, and also uses similar methods of comparing results of input that is in Chinese and in English.
I am adding it to this list though because I think it is interesting to see that these results of code being less accurate in other languages is seen as a repeated result is good to note, and speaks to how relevant this issue is with LLMs. 

Other mentioned works:
Ramel, D. (2024, February 5). Copilot by the numbers: Microsoft's big AI bet paying off. Visual Studio Magazine.[ https://visualstudiomagazine.com/Articles/2024/02/05/copilot-numbers.aspx](https://visualstudiomagazine.com/Articles/2024/02/05/copilot-numbers.aspx)
Discussion of co piolet usage.

### Overall:
Overall, the current literature brings to light some tools that could be useful for developing a generative LLM  based tool for wordplay, even with these tools the dataset that is available to train an LLM is most likely not going to be big enough to create skeletons of actual code, at least with the available tools right now. 
This is a field with a lot of resources and time being poured into it right now, so new tools that could help completely solve this data issue, might not be that far away. 
The literature also suggests that with some added parameters supporting translation, the margin of error with code skeletons could become fairly small, but there is very little research being done on generative models and translation on languages outside of Chinese, so being able to create accurate code skeletons in all the languages supported in Wordplay might not be viable, with the current available landscape of generate LLM tools. 
In my opinion, creating a tool that creates a generative LLM skeleton of code for Wordplay is a feature that is most likely only viable to create in the future, and not with the current tools and datasets available. 
Perhaps an alternative tool that could be useful is a code skeleton of comments to guide beginning users on how best to structure their program and what elements it might need. 
This would be much more viable as the skeleton would not be actual lines of code, so the margin for error  with generation would be much wider. 
It would also be more viable to adapt an already trained model for this task, as the logic behind how to structure programs is fairly equivalent whether you are using a language like Wordplay or python, for example, while actual code structure is much more specific to the actual programming language being used.

**Comment by amyjko** [comment]
This is an excellent summary of recent works and an outstanding foundation for design work on this issue. I especially like the way you synthesized this into a clear recommendation at the end. Thank you for your contribution to this issue!

**Comment by amyjko** [comment]
Once a quarter, I check each assignment to check the plans of the people that are assigned an issue. Please reply and indicate whether you intend to continue work on this issue. If you do not, please assign assign yourself, so we have an accurate account of who is working on an issue.

**Comment by amyjko** [comment]
No reply from @sdahl4, unassigning.

**Comment by amyjko** [comment]
Another possibility for project synthesis would be the prior research on *keyword programming*:

> Little, G., & Miller, R. C. (2007, November). Keyword programming in Java. In Proceedings of the 22nd IEEE/ACM International Conference on Automated Software Engineering (pp. 84-93).

It takes advantage of the smaller space of likely programs to help convert programming into a prompting interface. This may be simpler than LLMs, and more feasible, where we only have a small set of examples and no internet-scale training data.
